Seems reasonable, but there is a compilation failure:

```
Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
/home/vsts/work/1/s/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheFieldAccessEnhancer.java:[27,19] cannot find symbol
  symbol:   variable Gizmo
  location: class io.quarkus.panache.common.deployment.PanacheFieldAccessEnhancer.FieldAccessClassVisi
```

--- a/core/devmode/src/main/java/io/quarkus/dev/JavaCompilationProvider.java
+++ b/core/devmode/src/main/java/io/quarkus/dev/JavaCompilationProvider.java
@@ -21,7 +21,8 @@ import javax.tools.ToolProvider;
 
 import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;
-import org.objectweb.asm.Opcodes;
+
+import io.quarkus.gizmo.Gizmo;
 
 public class JavaCompilationProvider implements CompilationProvider {
 
@@ -88,7 +89,7 @@ public class JavaCompilationProvider implements CompilationProvider {
         private String sourceFile;
 
         public RuntimeUpdatesClassVisitor(Set<String> sourcePaths, String classesPath) {
-            super(Opcodes.ASM7);
+            super(Gizmo.ASM_API_VERSION);
             this.sourcePaths = sourcePaths;
             this.classesPath = classesPath;
         }

--- a/docs/src/main/asciidoc/writing-extensions.adoc
+++ b/docs/src/main/asciidoc/writing-extensions.adoc
@@ -743,7 +743,7 @@ void createExecutor(
         @Weak BuildProducer<GeneratedClassBuildItem> classConsumer,
         BuildProducer<ExecutorClassBuildItem> executorClassConsumer
 ) {
-        ClassWriter cw = new ClassWriter(Opcodes.ASM7);
+        ClassWriter cw = new ClassWriter(Gizmo.ASM_API_VERSION);
         String className = generateClassThatCreatesExecutor(cw); // <1>
         classConsumer.produce(new GeneratedClassBuildItem(true, className, cw.toByteArray()));
         executorClassConsumer.produce(new ExecutorClassBuildItem(className));

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/NoArgsConstructorProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/NoArgsConstructorProcessor.java
@@ -27,6 +27,7 @@ import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.gizmo.Gizmo;
 
 public class NoArgsConstructorProcessor {
 
@@ -91,7 +92,7 @@ public class NoArgsConstructorProcessor {
                         new BiFunction<String, ClassVisitor, ClassVisitor>() {
                             @Override
                             public ClassVisitor apply(String className, ClassVisitor classVisitor) {
-                                ClassVisitor cv = new ClassVisitor(Opcodes.ASM7, classVisitor) {
+                                ClassVisitor cv = new ClassVisitor(Gizmo.ASM_API_VERSION, classVisitor) {
 
                                     @Override
                                     public void visit(int version, int access, String name, String signature, String superName,

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertyBuildItemCandidateUtil.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertyBuildItemCandidateUtil.java
@@ -11,6 +11,8 @@ import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.Opcodes;
 
+import io.quarkus.gizmo.Gizmo;
+
 public class ConfigPropertyBuildItemCandidateUtil {
 
     /**
@@ -34,7 +36,7 @@ public class ConfigPropertyBuildItemCandidateUtil {
         private List<ConfigPropertyBuildItemCandidate> candidates;
 
         private ConfigClassVisitor(List<ConfigPropertyBuildItemCandidate> candidates) {
-            super(Opcodes.ASM7);
+            super(Gizmo.ASM_API_VERSION);
             this.candidates = candidates;
         }
 
@@ -54,7 +56,7 @@ public class ConfigPropertyBuildItemCandidateUtil {
 
         private ConfigClassConstructorVisitor(MethodVisitor superMethodVisitor,
                 List<ConfigPropertyBuildItemCandidate> candidates) {
-            super(Opcodes.ASM7, superMethodVisitor);
+            super(Gizmo.ASM_API_VERSION, superMethodVisitor);
             this.candidates = candidates;
         }
 

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
@@ -9,9 +9,9 @@ import org.hibernate.bytecode.spi.BytecodeProvider;
 import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.ClassWriter;
-import org.objectweb.asm.Opcodes;
 
 import io.quarkus.deployment.QuarkusClassWriter;
+import net.bytebuddy.utility.OpenedClassReader;
 
 /**
  * Used to transform bytecode by registering to
@@ -41,7 +41,7 @@ public final class HibernateEntityEnhancer implements BiFunction<String, ClassVi
         private final Enhancer enhancer;
 
         public HibernateEnhancingClassVisitor(String className, ClassVisitor outputClassVisitor) {
-            super(Opcodes.ASM7, new QuarkusClassWriter(ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS));
+            super(OpenedClassReader.ASM_API, new QuarkusClassWriter(ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS));
             this.className = className;
             this.outputClassVisitor = outputClassVisitor;
             //note that as getLoadingClassLoader is resolved immediately this can't be created until transform time

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/ProjectionForEnhancer.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/ProjectionForEnhancer.java
@@ -9,7 +9,8 @@ import org.objectweb.asm.AnnotationVisitor;
 import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.FieldVisitor;
 import org.objectweb.asm.MethodVisitor;
-import org.objectweb.asm.Opcodes;
+
+import io.quarkus.gizmo.Gizmo;
 
 public class ProjectionForEnhancer implements BiFunction<String, ClassVisitor, ClassVisitor> {
     private static final String BSONPROPERTY_BINARY_NAME = "org/bson/codecs/pojo/annotations/BsonProperty";
@@ -30,7 +31,7 @@ public class ProjectionForEnhancer implements BiFunction<String, ClassVisitor, C
         Map<String, String> propertyMapping;
 
         BsonPropertyClassVisitor(ClassVisitor outputClassVisitor, Map<String, String> propertyMapping) {
-            super(Opcodes.ASM7, outputClassVisitor);
+            super(Gizmo.ASM_API_VERSION, outputClassVisitor);
             this.propertyMapping = propertyMapping;
         }
 
@@ -38,7 +39,7 @@ public class ProjectionForEnhancer implements BiFunction<String, ClassVisitor, C
         public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
             FieldVisitor superVisitor = super.visitField(access, name, descriptor, signature, value);
             if (this.propertyMapping.containsKey(name)) {
-                return new FieldVisitor(Opcodes.ASM7, superVisitor) {
+                return new FieldVisitor(Gizmo.ASM_API_VERSION, superVisitor) {
                     private Set<String> descriptors = new HashSet<>();
 
                     @Override
@@ -65,7 +66,7 @@ public class ProjectionForEnhancer implements BiFunction<String, ClassVisitor, C
         public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
             MethodVisitor superVisitor = super.visitMethod(access, name, descriptor, signature, exceptions);
             if (this.propertyMapping.containsKey(name)) {
-                return new MethodVisitor(Opcodes.ASM7, superVisitor) {
+                return new MethodVisitor(Gizmo.ASM_API_VERSION, superVisitor) {
                     private Set<String> descriptors = new HashSet<>();
 
                     @Override

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheEntityEnhancer.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheEntityEnhancer.java
@@ -19,6 +19,7 @@ import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.Opcodes;
 import org.objectweb.asm.Type;
 
+import io.quarkus.gizmo.Gizmo;
 import io.quarkus.panache.common.Parameters;
 import io.quarkus.panache.common.Sort;
 import io.quarkus.panache.common.deployment.EntityField.EntityFieldAnnotation;
@@ -64,7 +65,7 @@ public abstract class PanacheEntityEnhancer<MetamodelType extends MetamodelInfo<
                 MetamodelInfo<? extends EntityModel<? extends EntityFieldType>> modelInfo,
                 ClassInfo panacheEntityBaseClassInfo,
                 ClassInfo entityInfo) {
-            super(Opcodes.ASM7, outputClassVisitor);
+            super(Gizmo.ASM_API_VERSION, outputClassVisitor);
             thisClass = Type.getType("L" + className.replace('.', '/') + ";");
             this.modelInfo = modelInfo;
             EntityModel<? extends EntityFieldType> entityModel = modelInfo.getEntityModel(className);
@@ -82,7 +83,7 @@ public abstract class PanacheEntityEnhancer<MetamodelType extends MetamodelInfo<
             }
             ef.signature = signature;
             // if we have a mapped field, let's add some annotations
-            return new FieldVisitor(Opcodes.ASM7, superVisitor) {
+            return new FieldVisitor(Gizmo.ASM_API_VERSION, superVisitor) {
                 private Set<String> descriptors = new HashSet<>();
 
                 @Override

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheFieldAccessEnhancer.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheFieldAccessEnhancer.java
@@ -4,7 +4,8 @@ import java.util.function.BiFunction;
 
 import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.MethodVisitor;
-import org.objectweb.asm.Opcodes;
+
+import io.quarkus.gizmo.Gizmo;
 
 public class PanacheFieldAccessEnhancer implements BiFunction<String, ClassVisitor, ClassVisitor> {
 
@@ -25,7 +26,7 @@ public class PanacheFieldAccessEnhancer implements BiFunction<String, ClassVisit
         private MetamodelInfo<?> modelInfo;
 
         public FieldAccessClassVisitor(String className, ClassVisitor outputClassVisitor, MetamodelInfo<?> modelInfo) {
-            super(Opcodes.ASM7, outputClassVisitor);
+            super(Gizmo.ASM_API_VERSION, outputClassVisitor);
             this.modelInfo = modelInfo;
             this.classBinaryName = className.replace('.', '/');
         }

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheFieldAccessMethodVisitor.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheFieldAccessMethodVisitor.java
@@ -4,6 +4,7 @@ import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.Opcodes;
 
 import io.quarkus.deployment.bean.JavaBeanUtil;
+import io.quarkus.gizmo.Gizmo;
 
 public class PanacheFieldAccessMethodVisitor extends MethodVisitor {
 
@@ -15,7 +16,7 @@ public class PanacheFieldAccessMethodVisitor extends MethodVisitor {
     public PanacheFieldAccessMethodVisitor(MethodVisitor methodVisitor, String owner,
             String methodName, String methodDescriptor,
             MetamodelInfo<?> modelInfo) {
-        super(Opcodes.ASM7, methodVisitor);
+        super(Gizmo.ASM_API_VERSION, methodVisitor);
         this.owner = owner;
         this.methodName = methodName;
         this.methodDescriptor = methodDescriptor;

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheMovingAnnotationVisitor.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheMovingAnnotationVisitor.java
@@ -1,8 +1,8 @@
 package io.quarkus.panache.common.deployment;
 
 import org.objectweb.asm.AnnotationVisitor;
-import org.objectweb.asm.Opcodes;
 
+import io.quarkus.gizmo.Gizmo;
 import io.quarkus.panache.common.deployment.EntityField.EntityFieldAnnotation;
 
 /**
@@ -14,7 +14,7 @@ public class PanacheMovingAnnotationVisitor extends AnnotationVisitor {
     private final EntityFieldAnnotation fieldAnno;
 
     public PanacheMovingAnnotationVisitor(EntityFieldAnnotation fieldAnno) {
-        super(Opcodes.ASM7);
+        super(Gizmo.ASM_API_VERSION);
         this.fieldAnno = fieldAnno;
     }
 

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java
@@ -16,6 +16,8 @@ import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.Opcodes;
 import org.objectweb.asm.Type;
 
+import io.quarkus.gizmo.Gizmo;
+
 public abstract class PanacheRepositoryEnhancer implements BiFunction<String, ClassVisitor, ClassVisitor> {
     private static final DotName OBJECT_DOT_NAME = DotName.createSimple(Object.class.getName());
 
@@ -42,7 +44,7 @@ public abstract class PanacheRepositoryEnhancer implements BiFunction<String, Cl
 
         public PanacheRepositoryClassVisitor(String className, ClassVisitor outputClassVisitor,
                 ClassInfo panacheRepositoryBaseClassInfo, IndexView indexView) {
-            super(Opcodes.ASM7, outputClassVisitor);
+            super(Gizmo.ASM_API_VERSION, outputClassVisitor);
             daoClassInfo = indexView.getClassByName(DotName.createSimple(className));
             daoBinaryName = className.replace('.', '/');
             this.panacheRepositoryBaseClassInfo = panacheRepositoryBaseClassInfo;

--- a/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
+++ b/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
@@ -61,6 +61,7 @@ import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.RuntimeInitializedClassBuildItem;
+import io.quarkus.gizmo.Gizmo;
 import io.quarkus.resteasy.common.deployment.JaxrsProvidersToRegisterBuildItem;
 import io.quarkus.resteasy.common.deployment.ResteasyCommonProcessor.ResteasyCommonConfig;
 import io.quarkus.resteasy.common.deployment.ResteasyDotNames;
@@ -560,7 +561,7 @@ public class ResteasyServerCommonProcessor {
                     .produce(new BytecodeTransformerBuildItem(name, new BiFunction<String, ClassVisitor, ClassVisitor>() {
                         @Override
                         public ClassVisitor apply(String className, ClassVisitor classVisitor) {
-                            ClassVisitor cv = new ClassVisitor(Opcodes.ASM7, classVisitor) {
+                            ClassVisitor cv = new ClassVisitor(Gizmo.ASM_API_VERSION, classVisitor) {
 
                                 @Override
                                 public void visit(int version, int access, String name, String signature, String superName,

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
@@ -3,6 +3,7 @@ package io.quarkus.arc.processor;
 import static io.quarkus.arc.processor.IndexClassLookupUtils.getClassByName;
 
 import io.quarkus.gizmo.DescriptorUtils;
+import io.quarkus.gizmo.Gizmo;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -163,7 +164,7 @@ final class Methods {
                     new BytecodeTransformer(classInfo.name().toString(), new BiFunction<String, ClassVisitor, ClassVisitor>() {
                         @Override
                         public ClassVisitor apply(String s, ClassVisitor classVisitor) {
-                            return new ClassVisitor(Opcodes.ASM7, classVisitor) {
+                            return new ClassVisitor(Gizmo.ASM_API_VERSION, classVisitor) {
                                 @Override
                                 public MethodVisitor visitMethod(int access, String name, String descriptor, String signature,
                                         String[] exceptions) {

--- a/integration-tests/class-transformer/deployment/src/main/java/io/quarkus/it/classtransformer/ClassTransformerProcessor.java
+++ b/integration-tests/class-transformer/deployment/src/main/java/io/quarkus/it/classtransformer/ClassTransformerProcessor.java
@@ -20,6 +20,7 @@ import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.gizmo.Gizmo;
 
 /**
  * class that adds an additional @GET @Path("/transformed") method to every JAX-RS endpoint.
@@ -52,7 +53,7 @@ public class ClassTransformerProcessor {
                 transformers.produce(new BytecodeTransformerBuildItem(i, new BiFunction<String, ClassVisitor, ClassVisitor>() {
                     @Override
                     public ClassVisitor apply(String className, ClassVisitor classVisitor) {
-                        ClassVisitor cv = new ClassVisitor(Opcodes.ASM7, classVisitor) {
+                        ClassVisitor cv = new ClassVisitor(Gizmo.ASM_API_VERSION, classVisitor) {
 
                             @Override
                             public void visit(int version, int access, String name, String signature, String superName,

